packages/ag-charts-community/src/components/toolbar/floatingToolbar.ts

```diff
@@ -0,0 +1,146 @@
+        this.dragged = false;
+        super.hide();
+    }
+
+    public getBounds() {
+        const element = this.getPopoverElement();
+        return {
+            x: element?.offsetLeft ?? 0,
+            y: element?.offsetTop ?? 0,
+            width: element?.offsetWidth ?? 0,
+            height: element?.offsetHeight ?? 0,
+        };
+    }
+
+    public hasBeenDragged() {
+        return this.dragged;
+    }
+
+    public setAnchor(anchor: ToolbarAnchor, horizontalSpacing: number, verticalSpacing: number) {
+        const element = this.getPopoverElement();
+        if (!element) return;
+
+        const position = anchor.position ?? 'above';
+        const { offsetWidth: width, offsetHeight: height } = element;
+
+        let top = anchor.y - height - verticalSpacing;
+        let left = anchor.x - width / 2;
+
+        if (position === 'below') {
+            top = anchor.y + verticalSpacing;
+        } else if (position === 'right') {
+            top = anchor.y - height / 2;
+            left = anchor.x + horizontalSpacing;
+        } else if (position === 'above-left') {
+            left = anchor.x;
+        }
+
+        this.updatePosition({ x: left, y: top });
+    }
+
+    public startDragging(event: MouseEvent, dragHandle?: HTMLElement) {
+        this.onDragStart(event, dragHandle);
+    }
+
+    public ignorePointerEvents() {
+        const element = this.getPopoverElement();
+        if (element) element.style.pointerEvents = 'none';
+    }
+
+    public capturePointerEvents() {
+        const element = this.getPopoverElement();
+        if (element) element.style.pointerEvents = 'unset';
+    }
+
+    protected override updatePosition(position: Vec2) {
+        const bounds = this.getBounds();
+
+        const canvasRect = this.ctx.domManager.getBoundingClientRect();
+        position.x = Math.floor(clamp(0, position.x, canvasRect.width - bounds.width));
+        position.y = Math.floor(clamp(0, position.y, canvasRect.height - bounds.height));
+
+        super.updatePosition(position);
+        this.onPopoverMoved();
+    }
+}
+
+export abstract class FloatingToolbar<
+    ButtonOptions extends ToolbarButtonOptions,
+    ButtonWidget extends ToolbarButtonWidget,
+> extends BaseToolbar<ButtonOptions, ButtonWidget> {
+    private readonly popover: FloatingToolbarPopover;
+
+    constructor(
+        ctx: ModuleContext,
+        id: string,
+        onButtonPress: (button: ButtonOptions & { index: number; sourceEvent: MouseEvent }) => void,
+        private readonly onToolbarMoved: (event: { buttonBounds: Array<BBoxValues>; popoverBounds: BBoxValues }) => void
+    ) {
+        super(ctx, onButtonPress, (event: MouseEvent, element: HTMLElement) =>
+            this.popover.startDragging(event, element)
+        );
+        this.popover = new FloatingToolbarPopover(ctx, id, this.onPopoverMoved.bind(this));
+    }
+
+    public show(options: PopoverOptions) {
+        this.popover.show([this.getElement()], options);
```

Comment:
This is acceptable for now. But eventually we'll want to change the signature of `FloatingToolbarPopover.show` to take a `Widget[]` rather a `HTMLElement[]`.

Reply:
Agreed, I'll look into converting popovers to widgets shortly.